File: native/android/src/main/kotlin/com/uulab/securekeypad/SecureKeypadPresentation.kt
package com.uulab.securekeypad

internal const val SECURE_KEYPAD_MAX_RENDERED_LENGTH = 4_096
internal const val SECURE_KEYPAD_ERROR_INTERNAL = 7
internal const val SECURE_KEYPAD_ERROR_INVALID = 1
internal const val SECURE_KEYPAD_NATIVE_REFRESH_ERROR = Long.MIN_VALUE

internal enum class SecureKeypadCommandDecision {
    ACCEPT,
    IGNORE,
    INVALID,
}

internal fun secureKeypadMonotonicCommandDecision(previous: Long?, requestId: Long): SecureKeypadCommandDecision {
    if (requestId < 0 || (previous != null && requestId < previous)) {
        return SecureKeypadCommandDecision.INVALID
    }
    if (previous != null && requestId == previous) {
        return SecureKeypadCommandDecision.IGNORE
    }
    return SecureKeypadCommandDecision.ACCEPT
}

/** Returns whether a native display-state code is part of the public contract. */
internal fun secureKeypadIsValidDisplayState(value: Int): Boolean = value in 0..3

/** Decodes the packed JNI state while keeping native failure distinct from empty state. */
internal fun secureKeypadDecodeMaskedState(packed: Long): Pair<Int, Int>? {
    if (packed == SECURE_KEYPAD_NATIVE_REFRESH_ERROR) return null
    return (packed ushr 32).toInt() to packed.toInt()
}

/** Converts a native display-state code without normalizing invalid values. */
internal fun secureKeypadDisplayStateName(value: Int): String = when (value) {
    0 -> "empty"
    1 -> "masked"
    2 -> "submitted"
    3 -> "cancelled"
    else -> "invalid"
}

/** Returns masked presentation text without ever accepting a user-supplied value. */
internal fun secureKeypadMaskedDisplayText(length: Int, protected: Boolean = false): String {
    require(length in 0..SECURE_KEYPAD_MAX_RENDERED_LENGTH)
    return if (protected) "Protected" else "•".repeat(length)
}

/** Returns an accessibility announcement containing only state and length. */
internal fun secureKeypadAccessibilityLabel(length: Int, protected: Boolean = false): String {
    require(length in 0..SECURE_KEYPAD_MAX_RENDERED_LENGTH)
    if (protected) return "Protected"
    return if (length == 0) "No input" else "$length characters entered"
}

/** Returns a deterministic security-facing presentation snapshot. */
internal fun secureKeypadSecuritySnapshot(length: Int, protected: Boolean = false, displayState: Int): String? {
    if (length !in 0..SECURE_KEYPAD_MAX_RENDERED_LENGTH || !secureKeypadIsValidDisplayState(displayState)) {
        return null
    }
    return listOf(
        "displayState=${secureKeypadDisplayStateName(displayState)}",
        "maskedDisplay=${secureKeypadMaskedDisplayText(length, protected)}",
        "accessibility=${secureKeypadAccessibilityLabel(length, protected)}",
        "protected=$protected",
    ).joinToString("\n")
}

File: packages/flutter/ios/Classes/SecureKeypadPresentation.swift
import Foundation

let secureKeypadMaxRenderedLength = 4_096
let secureKeypadInternalError: UInt32 = 7

/// Reorders only selected presentation items while preserving every row slot.
/// The caller owns the random source so production can require the platform
/// CSPRNG while contract tests can use a deterministic generator.
func secureKeypadPresentationRows<Item, Random: RandomNumberGenerator>(
    _ rows: [[Item]],
    randomizeInputKeys: Bool,
    isInput: (Item) -> Bool,
    using generator: inout Random
) -> [[Item]] {
    guard randomizeInputKeys else { return rows }
    var inputItems = rows.flatMap { $0 }.filter(isInput)
    guard inputItems.count > 1 else { return rows }
    for index in stride(from: inputItems.count - 1, through: 1, by: -1) {
        let swapIndex = Int.random(in: 0...index, using: &generator)
        inputItems.swapAt(index, swapIndex)
    }
    var inputIndex = 0
    return rows.map { row in
        row.map { item in
            guard isInput(item) else { return item }
            defer { inputIndex += 1 }
            return inputItems[inputIndex]
        }
    }
}

enum SecureKeypadCommandDecision: Equatable {
    case accept
    case ignore
    case invalid
}

func secureKeypadMonotonicCommandDecision(previous: Int64?, requestId: Int64) -> SecureKeypadCommandDecision {
    if requestId < 0 {
        return .invalid
    }
    if let previous, requestId < previous { return .invalid }
    if let previous, requestId == previous {
        return .ignore
    }
    return .accept
}

/// Returns whether the native presentation must hide sensitive UI state.
/// Screen capture remains protected across app-active transitions.
func secureKeypadShouldProtectPresentation(applicationIsActive: Bool, screenIsCaptured: Bool) -> Bool {
    !applicationIsActive || screenIsCaptured
}

/// Returns whether a native masked length is safe to render.
func secureKeypadIsValidRenderedLength(_ length: Int) -> Bool {
    (0...secureKeypadMaxRenderedLength).contains(length)
}

/// Returns whether a native display-state code is part of the public contract.
func secureKeypadIsValidDisplayState(_ value: UInt32) -> Bool {
    (0...3).contains(value)
}

/// Converts a native display-state code without normalizing invalid values.
func secureKeypadDisplayStateName(_ value: UInt32) -> String {
    switch value {
    case 0: return "empty"
    case 1: return "masked"
    case 2: return "submitted"
    case 3: return "cancelled"
    default: return "invalid"
    }
}

/// Produces masked display text from length only.
func secureKeypadMaskedDisplayText(length: Int, protected: Bool) -> String {
    precondition(secureKeypadIsValidRenderedLength(length))
    return protected ? "Protected" : String(repeating: "•", count: length)
}

/// Produces accessibility text from masked state only.
func secureKeypadAccessibilityLabel(length: Int, protected: Bool) -> String {
    precondition(secureKeypadIsValidRenderedLength(length))
    if protected {
        return "Protected"
    }
    return length == 0 ? "No input" : "\(length) characters entered"
}
